frame/offences/src/lib.rs

```diff
@@ -127,17 +150,42 @@ where
 		let slash_perbill: Vec<_> = (0..concurrent_offenders.len())
 			.map(|_| new_fraction.clone()).collect();
 
-		T::OnOffenceHandler::on_offence(
+		let applied = Self::report_or_store_offence(
 			&concurrent_offenders,
 			&slash_perbill,
 			offence.session_index(),
 		);
 
+		// Deposit the event.
+		Self::deposit_event(Event::Offence(O::ID, time_slot.encode(), applied));
+
 		Ok(())
 	}
 }
 
 impl<T: Trait> Module<T> {
+	/// Tries (without checking) to report an offence. Stores them in [`DeferredOffences`] in case
+	/// it fails. Returns false in case it has to store the offence.
+	fn report_or_store_offence(
+		concurrent_offenders: &[OffenceDetails<T::AccountId, T::IdentificationTuple>],
+		slash_perbill: &[Perbill],
+		session_index: SessionIndex,
+	) -> bool {
+		match T::OnOffenceHandler::on_offence(
+			&concurrent_offenders,
+			&slash_perbill,
+			session_index,
+		) {
+			Ok(_) => true,
+			Err(_) => {
+				<DeferredOffences<T>>::mutate(|d|
```

Comment:
You call this function in a loop, and each loop will call `mutate` on this storage item, pushing and fetching data from the overlay.

It would be better to keep track of a vector in memory, and do a single commit to deferred offenses if possible.

Reply:
Yeah good point, I will make an `//OPTIMIZATION` note for it to be tackled later. Although, Ain't overlay like the runtime cache and quite cheap to access? The first iteration is expensive and the rest stay in overlay, which is in-memory, so conceptually the same latency as what you proposed.

Reply:
I think memory overlay still does the trie stuff, so multiple overlay reads and multiple overlay writes each time you do this.

Im pretty sure that the cache is many times slower than in memory, but would be a good benchmark to test :)

Reply:
To be more precise, the change overlay (overlay_change.rs) is not running on a trie, but there is an additional cost associated with the overlay key value hashmap (currently btreemap) access.
So even if it can be considered small, I also think that running thing in memory when possible is way better :+1: .
The overlay can actually do some trie stuff (when calling ext_storage_root), but it recalculates the root at this time from the current value delta (ext_storage_root is very costy). Two consecutive call to ext_storage_root are cached, but if there is a single state change in between, everything is recalculated again. I understand it as a design choice where we favored storage operation over the root one.
Note, that for some case I think there can be multiple trie access: when producing proof we skip storage_cache so multiple read access to an unchanged value result in multiple trie access, but if the value got changed we still query the overlay change.

Reply:
a normal mutate doesn't call `ext_storage_root `, right? that should be called at the end of the block AFAIK? 

If it has storage root calculation it is probably more serious and deserves an issue, and we might make the same mistake in numerous places. 

If there are no trie/hashing/root stuff involved, as you said, is all in memory and relatively cheap and okay for now.

Reply:
No, it does not call storage root, I was only citing `ext_storage_root` because it is the only point when we have trie related processing during block processing.
It is indeed  processed at the end (unless we got a previous `ext_storage_root` and nothing written in between but that is just an optimization).
Actually I do not know which modules calls ext_storage_root, except contract for rent (and only the child variant of the storage root).

Reply:
@kianenigma In general, I had ideas about running tests over all of our pallets and extrinsics to try and identify this error. @pscott is working on https://github.com/paritytech/substrate/pull/5250 which will count the number of reads, writes, and overlay-reads.

This might help us identify when we make code which can be further optimized.

Reply:
> Yeah good point, I will make an `//OPTIMIZATION` note for it to be tackled later. Although, Ain't overlay like the runtime cache and quite cheap to access? The first iteration is expensive and the rest stay in overlay, which is in-memory, so conceptually the same latency as what you proposed.

Did this ever happen? I don't see it in the code.

Reply:
Yes it is fixed now.